File: health/src/lib/reportClientError.ts
// Reports client-side errors to the health dashboard's ingestion endpoint
// (reportClientError in health/functions/index.js -> Cloud Logging ->
// lib/clientErrors.js's collector -> findings on the Severity tabs). Works
// unmodified in Astro, Vite, plain React, or any other bundler -- it takes
// project/token as explicit arguments instead of reading a bundler-specific
// env var, so there is nothing here to adapt per framework.
//
// Copy this file into the app (e.g. src/lib/reportClientError.ts) and call
// initHealthReporting() once, as early as possible -- see this directory's
// README.md for exactly where, per framework.
//
// initHealthReporting() alone only catches what's *uncaught*:
// window.onerror / unhandledrejection. An error the app itself catches and
// handles -- shows the user a friendly "something went wrong" toast for,
// say -- never reaches either of those, so it would otherwise be invisible
// here no matter how often it happens. Call the exported reportError()
// directly from that catch block to put it on the dashboard too:
//
//   try {
//     await save(data)
//   } catch (err) {
//     showToast("Couldn't save -- please try again")
//     reportError(err instanceof Error ? err.message : String(err))
//   }
//
// Report the underlying technical message (err.message), not the friendly
// copy shown to the user -- that's what lets 500 occurrences of two
// different real causes show up as two distinct signatures on the
// dashboard instead of one uninformative "Couldn't save" bucket.
//
// Fire-and-forget by design: a failed report must never itself throw,
// retry, or block anything a real visitor is doing.
//
// Two independent layers guard against a page stuck in a render/retry loop
// (a useEffect with a bad dependency array is the recurring real-world
// shape), or a catch block on a hot path, turning into runaway cost:
//   1. Here: a per-tab session cap and a per-signature dedupe -- a single
//      tab can never send more than maxReportsPerSession requests, ever,
//      no matter how tight the loop.
//   2. health/functions/lib/rateLimit.js: a per-project, per-minute cap on
//      the endpoint itself -- the backstop for many tabs/devices hitting it
//      at once, which layer 1 alone can't see.
// Neither substitutes for the other -- keep both when copying this out.

export interface HealthReportingOptions {
  /** This app's id in health/functions/lib/config.js's PROJECTS list. */
  project: string;
  /**
   * This app's per-project secret from HEALTH_CLIENT_TOKENS (ask for one to
   * be minted when wiring up a new project -- see health-schema.md).
   * Pass undefined/"" for a build with no token configured (e.g. a fork or
   * local dev) -- reporting silently no-ops rather than sending bad auth.
   */
  token: string | undefined;
  /** Override only for local testing against an emulator. */
  endpoint?: string;
  /** Hard per-tab cap across the page's lifetime. Default 20. */
  maxReportsPerSession?: number;
}

export interface ReportExtra {
  stack?: string;
  /**
   * "warning" for a failure the app already handled gracefully (a caught
   * error behind a friendly message); "error" (the default) for anything
   * more serious. Purely informational -- both feed the same
   * clientErrorFloor/clientErrorCriticalFloor counts in config.js, so a
   * "warning" happening 500 times a day still escalates to critical.
   */
  level?: "error" | "warning";
}

const DEFAULT_ENDPOINT = "https://europe-west1-stanimeros-dev.cloudfunctions.net/reportClientError";
const DEFAULT_MAX_REPORTS_PER_SESSION = 20;

let initialized = false;
let state: { project: string; token: string; endpoint: string; maxReportsPerSession: number } | null = null;
const seen = new Set<string>();
let sent = 0;

/**
 * Wires up window.onerror / unhandledrejection reporting and enables
 * reportError() for manually-reported (caught) errors. Safe to call more
 * than once (e.g. React StrictMode double-invoking an effect, or a hot
 * reload) -- every call after the first is a no-op, so listeners are never
 * attached twice.
 */
export function initHealthReporting(options: HealthReportingOptions): void {
  if (initialized) return;
  initialized = true;

  const {
    project,
    token,
    endpoint = DEFAULT_ENDPOINT,
    maxReportsPerSession = DEFAULT_MAX_REPORTS_PER_SESSION,
  } = options;
  if (!token) return;

  state = { project, token, endpoint, maxReportsPerSession };

  window.addEventListener("error", (event) => {
    // A resource load failure (an <img>/<script> 404) fires "error" too,
    // but carries no `message` and no `error` -- nothing here for a report
    // to say.
    if (!event.message) return;
    reportError(event.message, { stack: event.error instanceof Error ? event.error.stack : undefined });
  });

  window.addEventListener("unhandledrejection", (event) => {
    const reason = event.reason as unknown;
    const message = reason instanceof Error ? reason.message : String(reason);
    reportError(message, { stack: reason instanceof Error ? reason.stack : undefined });
  });
}

/**
 * Report an error directly -- for a caught exception the app handles on its
 * own (shows a friendly message, falls back to a default, etc.) that would
 * otherwise never reach window.onerror/unhandledrejection. A no-op before
 * initHealthReporting() runs or when it no-op'd (no token configured).
 */
export function reportError(message: string | undefined, extra: ReportExtra = {}): void {
  if (!state || !message || sent >= state.maxReportsPerSession) return;
  const key = `${extra.level ?? "error"}:${message}`;
  if (seen.has(key)) return;
  seen.add(key);
  sent += 1;

  const body = JSON.stringify({
    project: state.project,
    token: state.token,
    message,
    stack: extra.stack,
    url: window.location.href,
    userAgent: navigator.userAgent,
    level: extra.level ?? "error",
  });

  try {
    // sendBeacon survives the page unloading mid-crash (a common moment for
    // an error to fire) without holding the navigation up; fall back to a
    // keepalive fetch where it isn't available. Its own failure is
    // swallowed -- there is nowhere left to report a failure to report an
    // error to.
    if (navigator.sendBeacon) {
      navigator.sendBeacon(state.endpoint, new Blob([body], { type: "application/json" }));
    } else {
      fetch(state.endpoint, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body,
        keepalive: true,
      }).catch(() => {});
    }
  } catch {
    // Swallowed on purpose -- see above.
  }
}

File: health/client-templates/flutter/report_client_error.dart
// Reports uncaught Flutter errors to the health dashboard's ingestion
// endpoint (reportClientError in health/functions/index.js -> Cloud
// Logging -> lib/clientErrors.js's collector -> findings on the Severity
// tabs). See ../web/reportClientError.ts for the web equivalent -- same
// contract, same protections, ported to Dart.
//
// Copy this file into the app (e.g. lib/report_client_error.dart) and call
// HealthReporting.init(project: ..., token: ...) as the first line of
// main(), wrapped in runZonedGuarded -- see this directory's README.md for
// the exact shape.
//
// Fire-and-forget by design, same as the web template: a failed report
// must never throw, retry, or crash the app a second time trying to report
// the first crash.
//
// Two independent layers guard against a rebuild loop (the Flutter
// analogue of a useEffect loop -- a widget stuck calling setState from
// build/didUpdateWidget) turning into runaway cost:
//   1. Here: a per-launch session cap and a per-signature dedupe -- one app
//      run can never send more than maxReportsPerSession requests, ever.
//   2. health/functions/lib/rateLimit.js: a per-project, per-minute cap on
//      the endpoint itself -- the backstop for many devices at once, which
//      layer 1 alone can't see.
// Neither substitutes for the other -- keep both when copying this out.
//
// Where Crashlytics is wired up (config.js's crashlyticsApps), it already
// covers native crashes and uncaught Dart errors on its own -- this exists
// for apps that haven't turned that BigQuery export on yet, or to fold an
// app's errors into the same dashboard as everything else's.

import 'dart:async';
import 'dart:convert';
import 'dart:ui';

import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;

class HealthReporting {
  HealthReporting._();

  static const _defaultEndpoint = 'https://europe-west1-stanimeros-dev.cloudfunctions.net/reportClientError';
  static const _defaultMaxReportsPerSession = 20;

  static bool _initialized = false;
  static String? _project;
  static String? _token;
  static String _endpoint = _defaultEndpoint;
  static int _maxReportsPerSession = _defaultMaxReportsPerSession;
  static final Set<String> _seen = {};
  static int _sent = 0;

  /// Call once, before runApp -- see this directory's README for where.
  /// Safe to call more than once (e.g. a hot restart) -- every call after
  /// the first is a no-op.
  static void init({
    required String project,
    required String? token,
    String endpoint = _defaultEndpoint,
    int maxReportsPerSession = _defaultMaxReportsPerSession,
  }) {
    if (_initialized) return;
    _initialized = true;
    // No token configured for this build (e.g. a fork or local dev) --
    // reporting silently no-ops rather than sending bad auth.
    if (token == null || token.isEmpty) return;

    _project = project;
    _token = token;
    _endpoint = endpoint;
    _maxReportsPerSession = maxReportsPerSession;

    final previousOnError = FlutterError.onError;
    FlutterError.onError = (FlutterErrorDetails details) {
      previousOnError?.call(details);
      report(details.exceptionAsString(), stack: details.stack?.toString());
    };

    // Catches what FlutterError.onError doesn't: errors thrown outside the
    // framework's own error zone (a bare async gap with no try/catch). The
    // runZonedGuarded call around runApp (see README.md) is what actually
    // routes those here -- PlatformDispatcher.instance.onError alone only
    // catches errors PlatformDispatcher itself originates.
    PlatformDispatcher.instance.onError = (Object error, StackTrace stack) {
      report(error.toString(), stack: stack.toString());
      return true; // handled -- don't crash release builds a second time over this
    };
  }

  /// Report an error directly (e.g. from a try/catch the app chose not to
  /// let propagate, but still wants on the dashboard).
  static void report(String? message, {String? stack, String level = 'error'}) {
    if (_project == null || _token == null) return;
    if (message == null || message.isEmpty || _sent >= _maxReportsPerSession) return;
    final key = '$level:$message';
    if (_seen.contains(key)) return;
    _seen.add(key);
    _sent += 1;

    // Never awaited by a caller, failures swallowed -- there is nowhere
    // left to report a failure to report an error to.
    unawaited(_send(message, stack, level));
  }

  static Future<void> _send(String message, String? stack, String level) async {
    try {
      await http
          .post(
            Uri.parse(_endpoint),
            headers: {'Content-Type': 'application/json'},
            body: jsonEncode({
              'project': _project,
              'token': _token,
              'message': message,
              'stack': stack,
              'level': level,
            }),
          )
          .timeout(const Duration(seconds: 5));
    } catch (_) {
      // Swallowed on purpose -- see class doc.
    }
  }
}
